### PR TITLE
Change jailer's 'unshare' + 'exec' to 'clone'

### PR DESCRIPTION
**Pull request by R4ZN1V**
The exec_into_new_pid_ns() function of the jailer should dispatch a new
process inside a new PID namespace.
It was done using unshare to a new PID namespace, then calling fork to
dispatch a process inside that namespace.
Changed it According to the TODO message - replaced this combo
with a single call to clone with the CLONE_NEWPID flag.

Besides the actual change inside the function, changed the jailer.md
documentation and the errors in main in order to fit the change.

# Reason for This PR

A TODO comment inside the code. Also it is a preferred linux way to use clone
in order to create a new child process when possible. 

## Description of Changes

 1. Changed the implementation of exec_into_new_pid_ns to use the clone function.
     Before the call I created a wrapper to give clone and a capture that will be the 
     Actual callback. The callback is passed to the wrapper using the arg parameter
     of the clone function.
     Also, mapped a stack for the new process using the mmap syscall.
 2. Changed the errors in main.rs - removed the UnshareNewPID and added MmapStack.
 3. Changed the documentation inside jailer.md to fit the change logic.

- [x] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: docs/jailer.md
Comment (luminitavoicu) [inline docs/jailer.md:51]:
```suggestion
  It makes use of the libc `clone()` function with the CLONE_NEWPID flag.
```

File: src/jailer/src/env.rs
Comment (luminitavoicu) [inline src/jailer/src/env.rs:217]: The `()` is not necessary here.

```suggestion
        let child_callback = || {
```

**Review by luminitavoicu — COMMENTED** [review]
Please sign your commit by adding `-s` to the git commit command. You can verify this by checking that you have a line that starts with: `Signed-off-by` in the commit message.

Also, it's useful to go through the `PR Checklist` and make sure all items that apply to your PR are addressed.

**Review by alindima — COMMENTED** [review]
Changes look Ok to me, but I see the pid-ns tests failing.
There are also some style tests failing.
Please also squash all your commits into one

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs:251]: So as you can see [here](https://github.com/lattera/glibc/blob/416bf844227d37b043b16be28c9523eeaecd3de3/nptl/sysdeps/unix/sysv/linux/x86_64/fork.c) libc's fork uses SIGCHLD when calling internally to the clone syscall. I wanted to minimize the changes as possible and to leave the same behavior as before (when the code used fork) - but if you would like I can remove this flag because the parent is definitely not waiting for this child.

**Review by serban300 — COMMENTED** [review]
Personally I have some concerns about this PR:

1. To me it looks like it adds complexity (the nested callbacks for example seem quite complex to me)
2. It adds a couple of unsafe blocks
3. The following code looks like double borrowing of `self`:
```
        let child_callback = || {
            // Reset process start time.
            self.start_time_cpu_us = 0;

            self.exec_command(chroot_exec_file.clone());
        };
        ...
        let child_pid = unsafe {
            libc::clone(
                child_fn,
                stack,
                libc::SIGCHLD | libc::CLONE_NEWPID,
                std::mem::transmute(Box::into_raw(Box::new(self))),
            )
        };
        self.save_exec_file_pid(child_pid, chroot_exec_file)?;
```

Are there any major advantages that we would get from taking this approach ? If not, personally I would keep the old one.

**Comment by luminitavoicu** [comment]
> Are there any major advantages that we would get from taking this approach ? If not, personally I would keep the old one.

`clone()` is faster than `unshare()` + `fork()` combination. As stated in the `TODO` comment, we would benefit from this change in case we needed to squeeze every drop of start-up latency we could get. However, I have never seen the start-up performance test failing since implementing this change.

File: src/jailer/src/env.rs
Comment (dianpopa) [inline src/jailer/src/env.rs:250]: The reason the tests fail is due to the fact that a SIGSEGV is thrown. The segfault is thrown when the child process starts using the stack provided by the parent process and it does so by growing down. The code needs to offer to clone the biggest address value of the stack:
```
let stack: &mut [u8; PROCESS_STACK_SIZE] = &mut [0; PROCESS_STACK_SIZE];
        let stack_ptr = unsafe { stack.as_mut_ptr().add(stack.len())
        };
...
let child_pid = unsafe {
            libc::clone(
                child_fn,
                stack_ptr as *mut libc::c_void,
                libc::SIGCHLD | libc::CLONE_NEWPID,
                std::mem::transmute(child_callback_ref),
            )
        };
```
This should make the tests pass.

File: src/jailer/src/env.rs
Comment (dianpopa) [inline src/jailer/src/env.rs:53]: nit: there is an extra space before "when"

File: docs/jailer.md
Comment (dianpopa) [inline docs/jailer.md:49]: There is also a reference to `new-pid-ns` here: https://github.com/firecracker-microvm/firecracker/blob/a39fabbe1ab39af41a154142858ce9e981414fa5/docs/jailer.md#jailer-operation. Please update that too.

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs:229]: Thought mmap is better for stack mapping than on the original thread + using the map_growsdown optimization, but it is better to avoid the extra syscall  for squeezing the delay - so I changed it to be a stack based buffer as you mentioned.

**Comment by dianpopa** [comment]
@R4ZN1V changes look good to me! 
There is still one test failing, the one ensuring the coverage does not drop. However, we cannot unit test the `exec_into_new_pid_ns` too easily since the function expects to be given the path to a valid firecracker binary. Feel free to modify the value of the [coverage limit for Intel](https://github.com/R4ZN1V/firecracker/blob/exec_into_new_pid_ns_to_clone/tests/integration_tests/build/test_coverage.py#L27) to 84.79.

**Review by dianpopa — APPROVED** [review]

**Comment by alindima** [comment]
Please rebase your branch on top of the latest main and fix the coverage conflict so that we can get this PR across the finish line

**Comment by R4ZN1V** [comment]
Hey - done the rebase. Lets get it across the finish line !

‫בתאריך יום ב׳, 9 באוג׳ 2021 ב-16:39 מאת ‪Alin Dima‬‏ <‪
***@***.***‬‏>:‬

**Review by dianpopa — DISMISSED** [review]

**Review by serban300 — CHANGES_REQUESTED** [review]
Sorry, as I mentioned before I have some doubts and I want to double check them. I'll try to do this ASAP. I'll block this PR a bit in the meanwhile.

**Comment by R4ZN1V** [comment]
Ok, Squashed the commits into one commit and fix the merge conflict. I think everything is Ok now and all the coverage test will pass. 
@dianpopa - Would you mind take a look at the coverage tests and approving again? sorry for the stale review, had to squash the commits. 
@serban300 - Ok, waiting for your review - If there is anything I can do to make it happen - let me now.
Thanks.

File: src/jailer/src/env.rs
Comment (serban300) [inline src/jailer/src/env.rs:210]: Nit: We don't need to declare the type here:

```
let callback_ref = arg as *mut *mut dyn std::ops::FnOnce();
```

File: src/jailer/src/env.rs
Comment (serban300) [inline src/jailer/src/env.rs:217]: Nit: For readability I would rename `child_fn` and `child_callback`. An option would be:
`child_fn` -> `on_clone`,
`child_callback` -> `do_on_clone`.

**Review by serban300 — CHANGES_REQUESTED** [review]

File: src/jailer/src/env.rs
Comment (alindima) [reply src/jailer/src/env.rs:231]: > It's not very clear to me what will happen next from rust's perspective. Are we sure that rust is not going to clear this memory when the parent process executes unsafe { libc::exit(0) } ? Would we need to use mem::forget() here to make sure ? Regardless, please also add a comment that explains this in more detail.

I've had this same concern. But after the `clone` call returns, the entire memory will have already been cloned to the new process (since it's not using `CLONE_VM `). So from the new process perspective, rust may do anything to the parent.

Comment (R4ZN1V) [reply src/jailer/src/env.rs:231]: @serban300 - It will not work. The libc clone wrapper forbids passing a NULL as stack (even if not using CLONE_VM flag). We can call directly to the syscall - but this will be super ugly in my opinion.

Comment (serban300) [reply src/jailer/src/env.rs:231]: Ok, I see. Then I'm ok with providing the stack.

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs]: I think the transmute is ok here, because there is no actual interpretation here, So it is safe. Also, your suggestion will not work and causes segfault.

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs]: So this will cause a segfault too, we have to let rust know it is a FnOnce instance.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs]: Sorry, I didn't have time to test the suggestions. It causes a segfault, because the cast in `on_clone()` has to be adapted to the new type. What seems to work is the following:

```
        let do_on_clone: &mut dyn FnOnce() = &mut || {
            // Reset process start time.
            self.start_time_cpu_us = 0;

            self.exec_command(chroot_exec_file.clone());
        };

        let do_on_clone_ptr = Box::into_raw(Box::new(do_on_clone));
```

and 

```
        extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
            let callback: Box<Box<dyn FnOnce()>> = unsafe { Box::from_raw(arg as _) };

            callback();
            0
        }
```

We need double pointers because trait objects are fat.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs]: Hmmm this is concerning because it would compile and then we would get an error at runtime. It might be easy to introduce bugs this way. 

Anyway, this problem, the double pointers problem, and the nested callbacks problem are all caused by the fact that we use the closure. Another option would be to capture the variables ourselves. Something like:

```
        struct Capture<'a> {
            env: &'a mut Env,
            chroot_exec_file: PathBuf,
        }

        let capture = Capture {
            env: self,
            chroot_exec_file,
        };
        let capture_ptr = Box::into_raw(Box::new(capture));
```

and keep only `on_clone`:

```
        extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
            let capture = unsafe { Box::<Capture>::from_raw(arg as _) };

            capture.env.exec_command(capture.chroot_exec_file.clone());

            0
        }
```

Not sure which option is best.

File: src/jailer/src/env.rs
Comment (serban300) [inline src/jailer/src/env.rs:237]: Should we handle the case where `child_pid` is < 0 ? This would mean that an error occurred.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs]: Also one more question comes to mind here. Sorry for insisting, but I am not very familiar with `clone()` and I would like to understand all the implications of this PR.

Whether we use a closure or we create a structure for passing arguments to the `on_clone()` callback, we end up passing a reference to self (For example the `env: &'a Env` in the previous comment). AFAIU If the `Env` struct was allocated on the parent's heap, we will find it in the child's process memory as well. But if the `Env` struct was allocated on the parent's stack, what will happen when `on_clone()` is executed ? What will the reference that we have in `Capture` point to ?

Comment (R4ZN1V) [reply src/jailer/src/env.rs]: So one very important thing you have to understand about clone - in our case (the flags I used) are really similar to the fork that was here before the commit. You can actually think about this change as a fork with the new-pid-ns flag on.
You can [look](https://github.com/lattera/glibc/blob/416bf844227d37b043b16be28c9523eeaecd3de3/nptl/sysdeps/unix/sysv/linux/x86_64/fork.c) on fork libc's source (that actually uses the fork syscall) and see it is exactly the same (except some thread flags that are really not important for us. 

The clone will actually copy all the memory to a new process just like fork does. Also, I think the capture is much more elegant and will be easier to understand than capturing all of them ourselves. There are also some disadvantages on capturing them ourselves BTW. I think all this part is ok and we can move on.

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs:235]: So I did not want to change the original behavior - originally the fork return value was not checked for error either. 
I thought it was done intentionally because then we only save the pid (-1 in an error case) and exit. 
I can change it to be different from the original behavior if you wish so.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs]: Ok, can you help me understand more in-depth these details ? Again, I don't know exactly how `clone()` works and I think these are crucial details, especially in the jailer where it can have security implications. I also started from the assumption that the way it is called, `clone()` should perform like `fork()`. But is this assumption true ? There are some corner cases that I can think of where I don't know how it would behave. So expanding on my previous question. What if the parent stack is 10k bytes and the `Env` or a captured variable is at the end of the parent's stack. Will it be copied ? If we used the original `clone()` syscall and we provide a null pointer as stack, my understanding is that indeed everything would have been copied identically. But what happens now since we allocate a stack of size 8192 bytes for the child ? What are the implications of doing this ?

> You can look on fork libc's source (that actually uses the clone syscall) and see it is exactly the same (except some thread flags that are really not important for us.

It's also providing a NULL stack, right ? 

> Also, I think the capture is much more elegant and will be easier to understand than capturing all of them ourselves. 

I agree. The capture looks better cosmetically.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs:235]: @luminitavoicu what do you think ? What was the reasoning behind this when implementing `fork()` ?

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs]: Ok, we shall look on the whole flow :)

1. We call libc's clone from the jailer.
2. Libc is [translating](https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/x86_64/clone.S.html
) to system call ABI and dispatching a clone syscall.
It is important to notice that the clone syscall gives different execution pathes acording to the return value (very similar to fork's behavior) and that libc translates it to a callback API for the user.
3. In the [kernel](https://code.woboq.org/linux/linux/kernel/fork.c.html#2327
), clone is defined to directly call _do_fork kernel function
4. Some unimportant checks for us, then calling [copy_process](https://code.woboq.org/linux/linux/kernel/fork.c.html#2226
)
5. A call to copy_mm is [made](https://code.woboq.org/linux/linux/kernel/fork.c.html#1916
), with the new process and the clone flags
6. We will take the [flow](https://code.woboq.org/linux/linux/kernel/fork.c.html#1374
) of CLONE_VM is not in the flags, which results in a deep copy of the mm of the original process
7. You can [see](https://code.woboq.org/linux/linux/kernel/fork.c.html#dup_mm
) that dup_mm actually allocates and memcpy the memory of the mm.

All of this will result in a FULL copy of the original process memory, including the original process stack :)

Comment (R4ZN1V) [reply src/jailer/src/env.rs]: Btw, the actual sp setting is done [here](https://code.woboq.org/linux/linux/arch/x86/kernel/process_64.c.html#420) inside copy_thread_tls which is called [here](https://code.woboq.org/linux/linux/kernel/fork.c.html#1925) inside copy_process.

Also, I made a boilerplate testing code that:
1. writes a value to its stack
2. call libc's clone
3. original father process exits
4. child process reads the value (after death) 

and everything is ok because as we said, this memory is copied. Thats it, I hope I shed some light for you. If anything is ok I think we finished everything except the return value of clone (@luminitavoicu ) and we just need to dispatch the tests in the buildkite and finally moving on :100:

Comment (serban300) [reply src/jailer/src/env.rs]: Thanks for the detailed explanations and links to kernel code. I will try to follow them in order to understand what the kernel does exactly. It will take some time. Anyway, yes, probably we're ok here. Just want to double check.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs:235]: Luminita is OOTO a couple of days so we'll probably have her input next week. Anyway personally I think we should return an Err when `child_pid` is < 0 . It would make it easier to spot a problem.

File: src/jailer/src/env.rs
Comment (serban300) [reply src/jailer/src/env.rs]: I looked on the kernel code and did some experiments. You're right, when calling clone the entire address space of the parent is copied and the parent's stack variables can be accessed at the same address as in the parent. The only problem is that as the child's stack grows it will eventually overwrite the previously allocated variables. For example in this sample code:

```
const PROCESS_STACK_SIZE: usize = 32;

fn main() {
    extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
        let data = arg as *mut u8;
        // Here (*data) == 37
        let stack_var_1 = 1u64;
        let stack_var_2 = 2u64;
        let stack_var_3 = 3u64;
        let stack_var_4 = 4u64;
        // Here (*data) == 4

        0
    }

    let mut data = 37u8;
    let data_ptr = &mut data as *mut u8;
    let arg: *mut libc::c_void = data_ptr as _;

    let mut stack = [0u8; PROCESS_STACK_SIZE];
    let stack_top_ptr = stack.last_mut().unwrap() as *mut u8;

    let child_pid = unsafe {
        libc::clone(
            on_clone,
            stack_top_ptr as _,
            libc::SIGCHLD | libc::CLONE_NEWPID,
            arg,
        )
    };

    if child_pid < 0 {
        println!("clone error: {}", std::io::Error::last_os_error());
    } else {
        println!("child_pid: {}", child_pid);
    }
}
```

You can see that we allocate `data = 37` and we pass it to the `on_clone()` callback. When we enter `on_call()` data is indeed 37, but as we allocate variables on the stack it is eventually overwritten and it becomes `4`. Not sure if this represents any reason for concern. I would just like to spend some time thinking about it. @firecracker-microvm/compute-capsule what do you think ?

Comment (serban300) [reply src/jailer/src/env.rs]: After spending some more time researching about this I think that the example above shows that it's hard to predict exactly where a variable will be allocated on the stack. That's because of compiler optimizations and the design of the code. Ideally we want the child's stack (`let mut stack = [0u8; PROCESS_STACK_SIZE];`) to be allocated at the bottom of the paren't stack, but we can't guarantee that with the current approach. Sometimes other variables might be allocated below it in which case we can end up overwriting them as the child's stack grows. This can be problematic if we actually need them in the child. I can think of 3 ways of allocating the child's stack:

1. As a slice allocated on the paren't stack. This is the current approach in the PR, and it doesn't look safe because of the reasons listed above.

2. On the heap with mmap. I'm not sure if this is ok since we don't know exactly where this memory is going to be mapped and if it's mapped really low in the address space, there might be little space for the stack and heap to grow. I'm not completely sure about this reasoning, but that's how it seems like. 

3. Using the clone syscall directly and providing a NULL stack pointer to it. I assume and hope that this option places the child's stack right at the bottom of the parent's stack which would be perfect. If this assumption is true, this seems like the safest option to me and I'm leaning towards it.

Please let me know if I'm missing anything.

Comment (R4ZN1V) [reply src/jailer/src/env.rs]: So:
2. If we will use mmap we can choose exactly the place of our mapping and place it where we would like - this will require extra logic and result in less clean code and more syscalls.
3. So this is not what will happen - if we will provide a NULL stack pointer, we will just continue with the original stack - and that's ok because as we talked, the VM_CLONE flag is not in use, and we will have a new replicated memory space - and this is exactly what fork do !

I will implement the raw syscall option, because we both understand this is the right option. I just want you to understand that there is no stack allocated in this case, because all the memory is copied and the virtual addresses will remain the same :)

Comment (serban300) [reply src/jailer/src/env.rs]: > 3. So this is not what will happen - if we will provide a NULL stack pointer, we will just continue with the original stack - and that's ok because as we talked, the VM_CLONE flag is not in use, and we will have a new replicated memory space - and this is exactly what fork do !

> I will implement the raw syscall option, because we both understand this is the right option. I just want you to understand that there is no stack allocated in this case, because all the memory is copied and the virtual addresses will remain the same :)

I see. It makes sense and indeed it seems the right option to me.

File: src/jailer/src/env.rs
Comment (luminitavoicu) [inline src/jailer/src/env.rs:228]: nit:
```suggestion
        // The call is safe because the stack size is a positive constant.
```

File: src/jailer/src/env.rs
Comment (luminitavoicu) [reply src/jailer/src/env.rs:235]: I am in favor of returning an error here if the child PID is negative, as it would make it easier to debug possible problems.

**Review by luminitavoicu — COMMENTED** [review]
PR looks good. Please add a check for the child PID, as agreed on https://github.com/firecracker-microvm/firecracker/pull/2652#discussion_r690984913. Also, rebase your PR and update x86_64 coverage, to make the build test pass.

File: src/jailer/src/env.rs
Comment (R4ZN1V) [reply src/jailer/src/env.rs:80]: @acatangiu Are you sure we want to unify the specific arch wrapper ?
I think its a bad habit, it will be super easy to insert a bug by changing one of the pid/ctid/new tls args.
I believe we should leave it that way, to prevent future bugs.
If you still insist, I will do it, I want to make that commit happen finally :)
@serban300 - can you approve the changes too in order to move on ? thanks.

Comment (serban300) [reply src/jailer/src/env.rs:80]: Personally I agree that it would be safer to use different logic for `x86_64` vs `aarch64`. Also the PR looks good.

**Review by serban300 — APPROVED** [review]

**Pull request merged**
